cxflow/hooks/confusion_matrix.py

```
@@ -0,0 +1,136 @@
+                 figure_action: str='save',
+                 normalize=True,
+                 cmap: str='Blues', **kwargs):
+        """
+        Create new :py:class:`SaveConfusionMatrix` hook.
+        :param output_dir: output directory
+        :param dataset: dataset (needed to translate predictions to strings)
+        :param annotations_name: annotation variable name
+        :param probabilities_name: prediction variable name
+        :param figure_action: action to be taken with the plotted figure, one of :py:attr:`FIGURE_ACTIONS`
+        :param normalize: False for plotting absolute values in confusion matrix, True for relative
+        :param cmap: type of colorbar  # http://matplotlib.org/examples/color/colormaps_reference.html
+        """
+        if figure_action not in SaveConfusionMatrix.FIGURE_ACTIONS:
+            raise ValueError('Unrecognized figure action `{}`. It must be one of `{}`'.
+                             format(figure_action, SaveConfusionMatrix.FIGURE_ACTIONS))
+
+        self._dataset = dataset
+        self._output_dir = output_dir
+        self._annotations_name = annotations_name
+        self._probabilities_name = probabilities_name
+        self._classes_name = classes_name
+        self._figure_action = figure_action
+        self._normalize = normalize
+        self._cmap = cmap
+
+        super().__init__(variables=list(filter(None, [annotations_name, probabilities_name, classes_name])), **kwargs)
+
+    def after_epoch(self, epoch_id: int, epoch_data: EpochData) -> None:
+        for stream_name, variables in self._accumulator.items():
+
+            probabilities = np.array(variables[self._probabilities_name])
+            predictions = np.argmax(probabilities, axis=-1)
+            annotations = np.array(variables[self._annotations_name])
+            # If annotations or have shape `num_data*num_classes`, take only the index of annotated class
+            if len(annotations.shape) == 2:
```

you can use `ndim` instead of `len(...shape)`

---

I'd check for higher ndims and raise an exception on unexpected formats of annotations